### PR TITLE
Add a cartopy.io.img_tiles.Stamen class

### PR DESCRIPTION
**Pull request by pelson**
## Rationale

Replaces #1013 to generalise the Stamen class of image tiles, thus giving access to all existing and future Stamen styles (following the same URL scheme).
Also adds a more abstract GoogleWTS tiler class to allow more appropriate subclassing.

## Implications

 * No further Stamen subclasses to be accepted (unless there is a new Stamen URL scheme)
 * All Tile classes should be updated to subclass GoogleWTS, not GoogleTiles (not done yet)
 * This is my review of #1013, and it closes #1013. cc @manugarri

### CODE REVIEW

File: lib/cartopy/io/img_tiles.py [hunk 1]
```diff
@@ -39,49 +39,14 @@
 import cartopy.crs as ccrs
 
 
-class GoogleTiles(object):
+class GoogleWTS(object):
```
Comment (QuLogic) [inline lib/cartopy/io/img_tiles.py]: Should this derive from `ABCMeta`? Our other base classes seem to do so.

File: lib/cartopy/io/img_tiles.py [hunk 2]
```diff
@@ -271,14 +294,14 @@ class StamenTerrain(GoogleTiles):
     https://github.com/migurski/DEM-Tools
 
     """
-    def _image_url(self, tile):
-        x, y, z = tile
-        url = 'http://tile.stamen.com/terrain-background/%s/%s/%s.png' % (
-            z, x, y)
-        return url
+    def __init__(self):
+        # NOTE: This subclass of Stamen exists for legacy reasons.
```
Comment (QuLogic) [inline lib/cartopy/io/img_tiles.py]: Deprecate it? Add note in docstring, too.

File: lib/cartopy/io/img_tiles.py [hunk 3]
```diff
@@ -244,15 +252,30 @@ def _image_url(self, tile):
...
         x, y, z = tile
         url = 'https://a.tile.openstreetmap.org/%s/%s/%s.png' % (z, x, y)
         return url
 
 
-class StamenTerrain(GoogleTiles):
+class Stamen(GoogleWTS):
+    """
+    Retrieves tiles from maps.stamen.com. Styles include
+    ``terrain-background``, ``terrain``, ``toner`` and ``watercolor``.
```
Comment (QuLogic) [inline lib/cartopy/io/img_tiles.py]: As noted in the other PR, it would be nice to explain what these are.

File: lib/cartopy/io/img_tiles.py (hunk 1 above)
Comment (pelson) [reply lib/cartopy/io/img_tiles.py]: ABC is pretty heavweight - personally I avoid it like the plague. Happy to implement if you feel strongly though.

File: lib/cartopy/io/img_tiles.py [hunk 4]
```diff
@@ -215,7 +170,60 @@ def get_image(self, tile):
...
+        style: optional
+            The style for the Google Maps tiles.  One of 'street',
+            'satellite', 'terrain', and 'only_streets'.  Defaults to 'street'.
+        url: optional
+            URL pointing to a tile source and containing {x}, {y}, and {z}.
+            Such as: ``'https://server.arcgisonline.com/ArcGIS/rest/services/\
+World_Shaded_Relief/MapServer/tile/{z}/{y}/{x}.jpg'``
+
+        """
+        # Only streets are partly transparent tiles that can be overlaid over
```
[the remark this answers is not in the document]
Comment (pelson) [reply lib/cartopy/io/img_tiles.py]: Frankly the comment is more like user documentation. I'm deleting it (happy for somebody else to actually turn this into compelling docs).

File: lib/cartopy/io/img_tiles.py (hunk 1 above)
Comment (pelson) [reply lib/cartopy/io/img_tiles.py]: **Re: our other base classes**

I can't argue with that. Consistency is definitely a thing. It is pretty galling for me though 😄

**Comment by pelson** [comment]
@QuLogic - updated as per review. Thanks for the feedback.

File: lib/cartopy/io/img_tiles.py [hunk 5]
```diff
@@ -265,20 +306,25 @@ class StamenTerrain(GoogleTiles):
     of flat, dark green.
 
     Additional info:
```
Comment (QuLogic) [inline lib/cartopy/io/img_tiles.py]: Maybe this should be changed to
```
References
----------
```
to match numpydoc style.

**Comment by QuLogic** [comment]
Looking at the warnings in the doc build, it seems like there might be a few other examples that use the `StamenTerrain` class.

**Comment by manugarri** [comment]
Oh nice!

File: lib/cartopy/io/img_tiles.py [hunk 6]
```diff
@@ -30,58 +30,25 @@
...
 import shapely.geometry as sgeom
 import numpy as np
 import six
-import warnings
 
 import cartopy.crs as ccrs
 
 
-class GoogleTiles(object):
+class GoogleWTS(six.with_metaclass(ABCMeta, object)):
```
Comment (ocefpaf) [inline lib/cartopy/io/img_tiles.py]: I love that this is getting the `WTMS` in the name but is it still worth keeping "Google" in there? Why not taking this opportunity to make the class name more generic? I believe `GoogleWTS` may throw users, who don't read the docs, off by thinking it only works for Google Tiles.

Comment (pelson) [reply lib/cartopy/io/img_tiles.py]: A fine observation. I agree entirely with the sentiment, but it is really hard to find a name that doesn't include the word Google - this isn't a general WMTS implementation AFAICT - it really does build upon Google mercator and uses (as far as I can tell) a non-standard y indexing scheme.

I suspect there are further abstractions that would be useful (there is a 'Microsoft' Quadtree implementation in here that really only needs a small subset of the GoogleWTS class's behaviour).

In short - I'm 100% in favour of finding a name that represents the scheme that Google implemented (I think it was them at least), but at present don't have a better suggestion.

**Comment by pelson** [comment]
@QuLogic  - merge at will. ``GoogleWTS`` rename is a thing I'd happily support in a subsequent PR / at a later release.
